Lombiq.OrchardCoreApiClient/ApiClient.cs

```diff
@@ -0,0 +1,93 @@
+using Lombiq.OrchardCoreApiClient.Constants;
+using Lombiq.OrchardCoreApiClient.Exceptions;
+using Lombiq.OrchardCoreApiClient.Interfaces;
+using Lombiq.OrchardCoreApiClient.Models;
+using RestEase;
+using System;
+using System.Collections.Generic;
+using System.Net.Http.Headers;
+using System.Threading.Tasks;
+
+namespace Lombiq.OrchardCoreApiClient;
+
+public class ApiClient
+{
+    private readonly ApiClientSettings _apiClientSettings;
+    private readonly IOrchardCoreApi _orchardCoreApi;
+
+    public ApiClient(ApiClientSettings apiClientSettings)
+    {
+        _apiClientSettings = apiClientSettings;
+        _orchardCoreApi = GetOrchardCoreApi(_apiClientSettings.DefaultTenantUri);
+    }
+
+    private IOrchardCoreApi GetOrchardCoreApi(Uri defaultTenantUri) =>
+        RestClient.For<IOrchardCoreApi>(defaultTenantUri, async (request, _) =>
+        {
+            if (request.Headers.Authorization != null)
+            {
+                var tokenResponse = await RestClient.
+                For<IOrchardCoreAuthorizatonApi>(defaultTenantUri).
+                TokenAsync(
+                    new Dictionary<string, string>
+                    {
+                        ["grant_type"] = "client_credentials",
+                        ["client_id"] = _apiClientSettings.ClientId,
+                        ["client_secret"] = _apiClientSettings.ClientSecret,
+                    });
+
+                tokenResponse.ResponseMessage.EnsureSuccessStatusCode();
+
+                if (tokenResponse.GetContent().Error != null)
+                {
+                    throw new ApiClientException(
+                        $"API client setup failed. An error occurred while retrieving an access token: " +
+                        tokenResponse.GetContent().Error);
+                }
+
+                if (tokenResponse.GetContent().ExpiresIn == "0")
+                {
+                    throw new ApiClientException($"The token is expired.");
+                }
```

Comment:
I don't know, but make sure whatever is needed actually happens. Keep in mind that this `ApiClient` may live for hours or days, and should continue to work. So if you need to do a new /token request before the next API request if the token expired (note my other remark about getting the token lazily, this should happen lazily too) then please do that.

Reply:
I think I finally understand fully how this works. In the [IOrchardCoreApi](https://github.com/Lombiq/Orchard-Core-API-Client/blob/c6b9630e69beb38f24f7ebbdd70b74f18f21b070/Lombiq.OrchardCoreApiClient/Interfaces/IOrchardCoreApi.cs) we need to authenticate to the endpoints, but when we authenticate with the client credentials, we get a new access token in the response, and this makes the expiration time reset. So actually I don't think we need to manually check for the expirations, since with every API call, we basically refresh the access token. I've tested it by setting up two tenants with a delay between them and logging the expiration times, and it was the same (3600 seconds is the default) every time.

Reply:
The token refresh is something that takes time, we don't want to do that for every single request. That's why instead you need to keep track of when the token expires (just having a field with a `DateTime` would suffice) and only issue this refresh request if necessary.